crates/turborepo-lib/src/shim.rs

```diff
@@ -176,50 +178,234 @@ struct PackageJson {
-            serde_json::from_reader(File::open(local_turbo_package_path).ok()?).ok()?;
+    // Unplugged strategy:
+    // - berry 2.1+
+    fn generate_unplugged_path(root_path: &Path) -> Option<PathBuf> {
+        let platform_package_name = TurboState::platform_package_name();
+        let unplugged_base_path = Self::get_unplugged_base_path(root_path);
+
+        unplugged_base_path
+            .read_dir()
+            .ok()
+            .and_then(|mut read_dir| {
+                // berry includes additional metadata in the filename.
+                // We actually have to find the platform package.
+                read_dir.find_map(|item| match item {
+                    Ok(entry) => {
+                        let file_name = entry.file_name();
+                        if file_name
+                            .to_string_lossy()
+                            .starts_with(platform_package_name)
+                        {
+                            Some(unplugged_base_path.join(file_name).join("node_modules"))
+                        } else {
+                            None
+                        }
+                    }
+                    Err(_) => None,
+                })
+            })
+    }
+
+    // We support six per-platform packages and one `turbo` package which handles
+    // indirection. We identify the per-platform package and execute the appropriate
+    // binary directly. We can choose to operate this aggressively because the
+    // _worst_ outcome is that we run global `turbo`.
+    //
+    // In spite of that, the only known unsupported local invocation is Yarn/Berry <
+    // 2.1 PnP
+    pub fn infer(root_path: &Path) -> Option<Self> {
+        let platform_package_name = TurboState::platform_package_name();
+        let binary_name = TurboState::binary_name();
+
+        let platform_package_json_path: PathBuf =
+            [platform_package_name, "package.json"].iter().collect();
+        let platform_package_executable_path: PathBuf =
+            [platform_package_name, "bin", binary_name].iter().collect();
+
+        // These are lazy because the last two are more expensive.
+        let search_functions = [
+            Self::generate_hoisted_path,
+            Self::generate_nested_path,
+            Self::generate_linked_path,
+            Self::generate_unplugged_path,
+        ];
+
+        // Detecting the package manager is more expensive than just doing an exhaustive
+        // search.
+        for root in search_functions
+            .iter()
+            .filter_map(|search_function| search_function(root_path))
+        {
+            // Needs borrow because of the loop.
+            #[allow(clippy::needless_borrow)]
+            let bin_path = root.join(&platform_package_executable_path);
+            match fs_canonicalize(&bin_path) {
+                Ok(bin_path) => {
+                    let resolved_package_json_path = root.join(platform_package_json_path);
```

Comment:
The early canonicalize is a performance thing since I need both `exists` and to grab the path for disambiguation. This gives me the same outcome as `.exists` with the side effect of returning the canonical value. `fs::read` handles it for us in the other case.